File: python/剑指offer/面试题59 - II. 队列的最大值.py
class MaxQueue:

    def __init__(self):
        from collections import deque
        self.queue = deque()
        self.max = deque()

    def max_value(self) -> int:
        if self.max:
            return self.max[0]
        else:
            return -1


    def push_back(self, value: int) -> None:
        self.queue.append(value)
        while self.max and value > self.max[-1]:
            self.max.pop()
        self.max.append(value)

    def pop_front(self) -> int:
        if self.queue:
            x = self.queue.popleft()
            if x == self.max[0]:
                self.max.popleft()
            return x
        else:
            return -1

File: go/剑指offer/面试题59 - II. 队列的最大值.go
package main

type MaxQueue struct {
	max []int
	queue []int
}


func Constructor() MaxQueue {
	return MaxQueue{
		max:   make([]int,0),
		queue: make([]int,0),
	}
}


func (this *MaxQueue) Max_value() int {
	if len(this.max) != 0{
		return this.max[0]
	}else{
		return -1
	}
}


func (this *MaxQueue) Push_back(value int)  {
	this.queue = append(this.queue,value)
	for len(this.max) != 0 && this.max[len(this.max)-1] < value{
		this.max = this.max[:len(this.max)-1]
	}
	this.max = append(this.max,value)
}


func (this *MaxQueue) Pop_front() int {
	if len(this.queue) == 0{
		return -1
	}
	x := this.queue[0]
	this.queue = this.queue[1:]
	if x == this.max[0]{
		this.max = this.max[1:]
	}
	return x
}
